core-plugins/src/main/java/co/cask/hydrator/plugin/batch/action/WindowsShareCopy.java

```diff
@@ -0,0 +1,276 @@
+    StringBuilder sb = new StringBuilder("smb://");
+    sb.append(config.netBiosHostname);
+    sb.append("/");
+    sb.append(config.netBiosSharename);
+    sb.append("/");
+    if (config.sourceDirectory.startsWith("/")) {
+      if (config.sourceDirectory.length() > 1) {
+        sb.append(config.sourceDirectory.substring(1));
+      }
+    } else {
+      sb.append(config.sourceDirectory);
+    }
+    final String smbDirectory = sb.toString();
+
+    // Register the SMB File handler.
+    jcifs.Config.registerSmbURLHandler();
+    jcifs.Config.setProperty("jcifs.util.loglevel", "4");
+
+    // Authentication with NTLM and read the directory from the Windows Share.
+    final NtlmPasswordAuthentication auth = new NtlmPasswordAuthentication(config.netBiosDomainName,
+                                                                           config.netBiosUsername,
+                                                                           config.netBiosPassword);
+    SmbFile dir = new SmbFile(smbDirectory, auth);
+
+    // Create the HDFS directory if it doesn't exist.
+    final Path hdfsDir = new Path(config.destinationDirectory);
+    final FileSystem hdfs = hdfsDir.getFileSystem(new Configuration());
+    if (!hdfs.exists(hdfsDir)) {
+      hdfs.mkdirs(hdfsDir);
+    }
+
+    // Determines the buffer size to be used during writing.
+    Configuration conf = new Configuration();
+    conf.setLong("io.file.buffer.size", config.bufferSize);
+    String[] list = dir.list();
+    int threadCount = (config.numThreads == null) ? 1 : config.numThreads;
+
+    // Copies the files in a multithreaded way
+    CountDownLatch executorTerminateLatch = new CountDownLatch(1);
+    ExecutorService executorService = createExecutor(threadCount, executorTerminateLatch);
+    CompletionService<String> completionService = new ExecutorCompletionService<>(executorService);
+
+    try {
+      for (final String file : list) {
+        completionService.submit(new Callable<String>() {
+          @Override
+          public String call() throws Exception {
+            String fileName = "";
+            try {
+              if (smbDirectory.endsWith("/")) {
+                fileName = copyFileToHDFS(hdfs, smbDirectory + file, hdfsDir, auth);
+              } else {
+                fileName = copyFileToHDFS(hdfs, smbDirectory + "/" + file, hdfsDir, auth);
+              }
+            } catch (IOException e) {
+              LOG.warn("Exception while copying the file {}", file, e);
+            }
+            return fileName;
+          }
+        });
+      }
+
+      for (int i = 0; i < list.length; i++) {
+        try {
+          Future<String> fileWritten = completionService.take();
+          String fileName = fileWritten.get();
+          LOG.info("{} is copied", fileName);
+        } catch (Throwable t) {
+          Throwable rootCause = Throwables.getRootCause(t);
+          if (rootCause instanceof ExecutionException) {
```

Comment:
I don't find much value in catching the exception, and then logging+rethrowing the root cause.
In fact logging and rethrowing the exception result in redundant error messages (since the thrown exception will be caught and logged at a higher level).